File: Array_problems/ReshapeTheMatrix.js

// Problem link - https://leetcode.com/problems/reshape-the-matrix/?envType=problem-list-v2&envId=array

// Approach - 1 (Brute Force)
let matrixReshape1 = function(mat, r, c) {
    // find row and column size of input matrix `mat`
    const m = mat.length;
    const n = mat[0].length;

    // step 1 - check for edge case : check if reshap is possible or not
    if(m * n !== r * c){
        // if reshape is not possible then return original input matrix `mat`
        return mat;
    }

    // step 2 - flatten the original input 2-D array `mat` into 1-D array `flatten`
    // this is done to maintain original array order
    const flatten = []
    for(let i=0; i<m; i++){
        for(let j=0; j<n; j++){
            flatten.push(mat[i][j]);
        }
    }

    // store final result in reshaped matrix `reshape`
    const reshape = [];
    
    // keep track of index
    let index = 0;

    // step 3 - fill the new matrix row by row using flattened data
    for(let i=0; i<r; i++){
        // create new array
        const row = [];
        for(let j=0; j<c; j++){
            // add element to current row
            row.push(flatten[index]);
            // move to next element in flatten array
            index++;
        }

        // add completed row to reshaped matrix
        reshape.push(row);
    }

    // step 4 - return reshaped matrix `reshaped`
    return reshape;
};

// Approach2 - Optimal
let matrixReshape = function(mat, r, c) {
    // find row and column size of input matrix `mat`
    const m = mat.length;
    const n = mat[0].length;

    // step 1 - check for edge case : if total element differ then reshap is not possible
    if(m * n != r * c){
        return mat;
    }

    // store final result in `reshaped` matrix
    const reshaped = Array.from({ length: r }, () => Array(c).fill(0));

    // step 2 - fill the new `reshaped` matrix 
    // single linear pass over all elements (row-major order)
    for(let k=0; k<m*n; k++){
        // step 3 - calculate original position in matrix `mat`
        let originalRow = Math.floor(k / n);
        let originalColumn = k % n;

        // step 4 - find new position in reshaped matrix
        let newRow = Math.floor(k / c);
        let newColumn = k % c;

        // move the element
        reshaped[newRow][newColumn] = mat[originalRow][originalColumn];
    } 
        
    // step 5 - return reshaped matrix `reshaped`
    return reshaped;
};

const mat = [[1,2],[3,4]], r = 1, c = 4;

// const mat = [[1,2],[3,4]], r = 2, c = 4;
console.log(matrixReshape(mat, r, c));

File: Array_problems/ReshapeTheMatrix.cpp

// Problem link - https://leetcode.com/problems/reshape-the-matrix/description/?envType=problem-list-v2&envId=array

#include<bits/stdc++.h>
using namespace std;

vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
    // find row and column size of input matrix `mat`
    int m = mat.size();
    int n = mat[0].size();

    // step 1 - check for edge case : if total element differ then reshap is not possible
    if (m * n != r * c) {
        return mat;
    }

    // store final result in `reshaped` matrix
    vector<vector<int>> reshaped(r, vector<int>(c));

    // step 2 - fill the new `reshaped` matrix 
    // single linear pass over all elements (row-major order)
    for (int k = 0; k < m * n; k++) {
        // step 3 - calculate original position in matrix `mat`
        int originalRow = k / n;
        int originalColumn = k % n;

        // step 4 - find new position in reshaped matrix
        int newRow = k / c;
        int newColumn = k % c;

        // move the element
        reshaped[newRow][newColumn] = mat[originalRow][originalColumn];
    }

    // step 5 - return reshaped matrix `reshaped`
    return reshaped;
}

void print2DVector(vector<vector<int>>& matrix){
    cout << "[";
    for (size_t i = 0; i < matrix.size(); ++i) {
        cout << "[";
        for (size_t j = 0; j < matrix[i].size(); ++j) {
            cout << matrix[i][j];
            if (j < matrix[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < matrix.size() - 1) cout << ",";
    }
    cout << "]" << std::endl;
}

int main(){
    // vector<vector<int>> mat = {{1,2},{3,4}};
    // int r = 1, c = 4;

    vector<vector<int>> mat = {{1,2},{3,4}};
    int r = 2, c = 4;
    vector<vector<int>> res = matrixReshape(mat, r, c);
    print2DVector(res);
    return 0;
}